src/Hamiltonians/abstract.jl

```
@@ -298,16 +298,177 @@ struct BasisSetRep{A,SM,H}
+    # symmetrisation of the sparse matrix
+    return _bsr_ensure_symmetry(AdjointUnknown(), h, addr; kwargs...)
+end
+# special cases are needed for symmetry wrappers and are defined there
+
+# default, does not enforce symmetries
+function _bsr_ensure_symmetry(
+    ::LOStructure, h::AbstractHamiltonian, addr;
+    sizelim=10^6, test_approx_symmetry=true, kwargs...
+)
+    dimension(Float64, h) < sizelim || throw(ArgumentError("dimension larger than sizelim"))
+    check_address_type(h, addr)
+    sm, basis = build_sparse_matrix_from_LO(h, addr; kwargs...)
+    return BasisSetRep(sm, basis, h)
+end
+
+# build the BasisSetRep while enforcing hermitian symmetry
+function _bsr_ensure_symmetry(
+    ::IsHermitian, h::AbstractHamiltonian, addr;
+    sizelim=10^6, test_approx_symmetry=true, kwargs...
 )
     dimension(Float64, h) < sizelim || throw(ArgumentError("dimension larger than sizelim"))
     check_address_type(h, addr)
     sm, basis = build_sparse_matrix_from_LO(h, addr; kwargs...)
+    fix_approx_hermitian!(sm; test_approx_symmetry) # enforce hermitian symmetry after building
     return BasisSetRep(sm, basis, h)
 end
 
+"""
+    fix_approx_hermitian!(A; test_approx_symmetry=true, kwargs...)
+Replaces the matrix `A` by `½(A + A')` in place. This will be successful and the result
+is guaranteed to pass the `ishermitian` test only if the matrix is square and already
+approximately hermitian.
+
+By default logs an error message if the matrix `A` is found to not be approximately
+hermitian. Set `test_approx_symmetry=false` to bypass testing.
+Other keyword arguments are passed on to `isapprox`.
+"""
+function fix_approx_hermitian!(A; test_approx_symmetry=true, kwargs...)
+    # Generic and inefficient version. Make sure to replace by efficient specialised code.
+    if test_approx_symmetry
+        passed = isapprox(A, A'; kwargs...)
+        if !passed
+            @error "Matrix is not approximately hermitian."
```

Should we throw an error here?

---

Done.